polly/src/pollable.rs

```diff
@@ -0,0 +1,197 @@
+        Pollable { fd }
+    }
+}
+
+impl AsRawFd for Pollable {
+    fn as_raw_fd(&self) -> RawFd {
+        self.fd
+    }
+}
+
+pub type EventRegistrationData = (Pollable, EventSet);
+
+#[derive(PartialEq)]
+pub enum PollableOp {
+    /// Register a new handler for a pollable and eventset.
+    Register(EventRegistrationData),
+    /// Unregister a handler for a pollable.
+    Unregister(EventRegistrationData),
+    /// Update eventset for a specified pollable.
+    Update(EventRegistrationData),
+}
+
+impl std::fmt::Debug for PollableOp {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        use self::PollableOp::*;
+
+        match self {
+            Register(data) => write!(f, "Register {:?}", data),
+            Unregister(data) => write!(f, "Unregister {:?}", data),
+            Update(data) => write!(f, "Update {:?}", data),
+        }
+    }
+}
+
+bitflags! {
+    #[derive(Default)]
+    pub struct EventSet: u8 {
+        const NONE = 0b0000_0000;
+        const READ = 0b0000_0001;
+        const WRITE = 0b0000_0010;
+        const CLOSE = 0b0000_0100;
+    }
+}
+
+impl EventSet {
+    /// Read event.
+    pub fn readable(self) -> bool {
+        self.contains(EventSet::READ)
+    }
+    /// Write event.
+    pub fn writeable(self) -> bool {
+        self.contains(EventSet::WRITE)
+    }
+    // Close event.
+    pub fn closed(self) -> bool {
+        self.contains(EventSet::CLOSE)
+    }
+}
+
+impl std::fmt::Display for Pollable {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        write!(f, "{}", self.fd)
+    }
+}
+
+pub struct PollableOpBuilder {
+    fd: Pollable,
+    event_mask: EventSet,
+}
+
+impl PollableOpBuilder {
+    /// Constructs a new PollableOp builder for the specified Pollable.
+    pub fn new(fd: Pollable) -> PollableOpBuilder {
+        PollableOpBuilder {
+            fd,
+            event_mask: EventSet::NONE,
+        }
+    }
+
+    /// Caller is interested in Pollable read events.
+    pub fn readable(&mut self) -> &mut PollableOpBuilder {
+        self.event_mask |= EventSet::READ;
+        self
+    }
+
+    /// Caller is interested in Pollable write events.
+    pub fn writeable(&mut self) -> &mut PollableOpBuilder {
+        self.event_mask |= EventSet::WRITE;
+        self
+    }
+
+    /// Caller is interested in Pollable close events.
+    pub fn closeable(&mut self) -> &mut PollableOpBuilder {
+        self.event_mask |= EventSet::CLOSE;
+        self
+    }
+
+    /// Create a Register PollableOp.
+    pub fn register(&self) -> PollableOp {
+        PollableOp::Register((self.fd.clone(), self.event_mask))
+    }
+
+    /// Create an Unregister PollableOp.
+    pub fn unregister(&self) -> PollableOp {
+        PollableOp::Unregister((self.fd.clone(), self.event_mask))
+    }
+
+    /// Create an Update PollableOp.
+    pub fn update(&self) -> PollableOp {
+        PollableOp::Update((self.fd.clone(), self.event_mask))
+    }
```

Comment:
Do these two operations make sense in the context of `PollableOpBuilder`? In other words, do we expect to have the following line of code as a valid usecase:
```rust
let pollable = PollableOpBuilder::new(some_pollable_obj).writable().readable().unregister()
```

With the builder pattern in mind, I was picturing the PollableOpBuilder to only have configuration functions (such as writable, readable, closable) and the register function which returns the built object (in this case PollableOp). The `update` and `unregister` operations I was expecting to see in PollableOp or some other place.

Reply:
Yes, all 3 operations register/unregister/update are possible.

Let's think about the scenario where we are registering and waiting for only read events:
PollableOpBuilder::new(some_pollable_obj).readable().register()

For example an EventHandler implementer would want to wait for the write event by returning 
this pollableop from one of its handle_XXX callbacks:
PollableOpBuilder::new(some_pollable_obj).readable().writeable().update()

Reply:
My confusion came from naming. I was expecting PollableOp to mean operation in the sense of something being executed. This enum is more like a type of pollable operation in which case it makes sense for the builder to have 3 methods that return all possible types of operations. I think we should change the name of the enum & builder to better reflect what they represent.